packages/neon_http_client/lib/src/interceptors/cookie_interceptor.dart

```diff
@@ -0,0 +1,98 @@
+import 'dart:async';
+
+import 'package:cookie_store/cookie_store.dart';
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+import 'package:neon_http_client/src/interceptors/http_interceptor.dart';
+import 'package:universal_io/io.dart';
+
+/// A HttpInterceptor to implement cookie persisting interceptors.
+///
+/// If the request already contains a 'cookie' header the request will not be intercepted.
+@immutable
+abstract class CookieInterceptor implements HttpInterceptor {
+  /// Creates a new cookie interceptor.
+  const CookieInterceptor();
+
+  @override
+  bool shouldInterceptRequest(http.BaseRequest request) {
+    return !request.headers.containsKey(HttpHeaders.cookieHeader);
+  }
+
+  @override
+  Future<http.BaseRequest> interceptRequest({required http.BaseRequest request}) async {
+    if (!shouldInterceptRequest(request)) {
+      throw StateError('Request already contains a cookie header and should not be intercepted.');
+    }
```

Comment:
Checking this again feels useless to me, the caller should be responsible for only calling when allowed

Reply:
> `should`

I'd rather have a sanity check to catch any wrongdoing. Should i can change it to an assertion?

Reply:
Yeah assertion is good